src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -849,25 +873,106 @@ static MemorySegment ofNativeRestricted() {
+     * explicitly permitted subtypes.
+     *
+     * @implSpec
+     * Implementations of this interface are not thread-safe.
+     *
+     * @see MemorySegment#handoff(HandoffTransform)
      */
-    int ALL_ACCESS = READ | WRITE | CLOSE | SHARE | HANDOFF;
+    interface HandoffTransform {
+
+        /**
+         * Creates a new shared handoff transform. The target segment produced by the returned transform is a shared
+         * segment, which can be accessible concurrently from multiple threads.
+         * @return a shared handoff transform.
+         */
+        static HandoffTransform ofShared() {
+            return new AbstractMemorySegmentImpl.HandoffTransformImpl(null);
+        }
+
+        /**
+         * Creates a new confined handoff transform, bound by a given thread. The target segment produced by the returned transform
+         * is a confinement segment, which will only be accessible to a specific thread.
+         *
+         * @param t the confinement thread.
+         * @return a confined handoff transform.
+         * @throws NullPointerException if {@code t == null}.
+         */
+        static HandoffTransform ofConfined(Thread t) {
+            return new AbstractMemorySegmentImpl.HandoffTransformImpl(t);
+        }
+
+        /**
+         * Creates a new confined handoff transform, bound by the current thread (see {@link Thread#currentThread()}).
+         * The target segment produced by the returned transform is a confinement segment, which will only be accessible to a specific thread.
+         *
+         * @return a confined handoff transform.
+         * @throws NullPointerException if {@code t == null}.
+         */
+        static HandoffTransform ofConfined() {
+            return ofConfined(Thread.currentThread());
+        }
+
+        /**
+         * Specifies an additional cleanup action to be associated with the target segment. More specifically, the target
+         * segment produced by this transform will feature a cleanup action which will first call the user-provided action,
+         * before delegating back to cleanup action associated with the source segment.
+         * Any errors and/or exceptions thrown by the user-provided action will be discarded, and will not prevent the
+         * release of any memory resources associated with the target segment.
+         *
+         * @param action the new cleanup action
+         * @return this handoff transform.
+         * @throws NullPointerException if {@code action == null}
+         */
+        HandoffTransform addCleanupAction(Runnable action);
+
+        /**
+         * Specifies an additional attachment object which will be kept alive by the target segment, in addition
+         * to any attachment object which might also be associated with the source segment.
+         * This can be useful in cases where the lifecycle of the segment is dependent on that of some other external resource.
```

Comment:
I guess what I wanted to mean here is if the segment depends on some memory source which is maintained externally and must be kept alive. That is the bytebuffer case - you have a segment which has to keep the BB alive, otherwise access with segfault.